src/Datadog.Trace/Containers/ContainerInfo.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text.RegularExpressions;
+using System.Threading;
+
+namespace Datadog.Trace.Containers
+{
+    /// <summary>
+    /// Utility class with methods to interact with container hosts.
+    /// </summary>
+    internal static class ContainerInfo
+    {
+        private const string ControlGroupsFilePath = "/proc/self/cgroup";
+
+        private const string ContainerIdRegex = @"^(?:\d+):(?:[^:]*):/?(?:.+/)([0-9a-f]{8}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{12}|[0-9a-f]{64}(?:\.scope)?)$";
+
+        private static readonly Lazy<string> ContainerId = new Lazy<string>(GetContainerIdInternal, LazyThreadSafetyMode.ExecutionAndPublication);
+
+        /// <summary>
+        /// Gets the id of the container executing the code.
+        /// Return <c>null</c> if code is not executing inside a supported container.
+        /// </summary>
+        /// <returns>The container id or <c>null</c>.</returns>
+        public static string GetContainerId()
+        {
+            return ContainerId.Value;
```

Comment:
Is it worth wrapping the code to retrieve the Id in a try catch?
Might be a good policy for new features which are cross cutting like this.
It would make sure that in any catastrophic miss, we don't stop everything else from working, we just don't get the extra data.

Reply:
Could log about it and maybe even set the container id to some reserved value that we could treat as an error state.
If a customer ever says, "Hey, my container id isn't populated", we could differentiate between a null and some specific value which we know indicates the code bombed.

Reply:
I'm not sure if the try/catch should be so high up. Since the failures would only happen with the I/O, I'm suggesting in a different comment that we wrap the specific calls with try/catch

Reply:
I added exception handling and logging in d78a4cc1f7a92714e4c4aabef45c5dd7a086c4f5.